🐛 Fixed settings overriden when updated from multiple tabs

jbenezech opened the pull request:
closes #15470
When multiple browser tabs are open, each manipulate a different copy of ember data model, changes to the model in one tab are not reflected in the model of the other tab. When updating some settings, all current settings were sent to the API. As a result, when updating two different categories of settings (navigation/code inspection) in different tabs, the second update was overriding the first one. From a user perspective, this is not a natural behaviour. Only settings visible on-screen when clicking save should be modified.

jbenezech committed:
:bug:Fixed settings overriden when updated from multiple tabs

closes #15470
When multiple browser tabs are open, each manipulate a different copy of ember data model, changes to the model in one tab are not reflected in the model of the other tab.
When updating some settings, all current settings were sent to the API.
As a result, when updating two different categories of settings (navigation/code inspection) in different tabs, the second update was overriding the first one.
From a user perspective, this is not a natural behaviour. Only settings visible on-screen when clicking save should be modified.

jbenezech commented:
This PR should get the job done but is not ideal.  

Relying on change detection is risky as it will not work on non-trivial model attributes (such as NavigationItem). The workaround put in place (force a change in the controller) will have to be done for new similar settings. Failure to do so will not be detected in the tests and the issue will come back).  
But there are no "easy" solution I think
- Syncing ember data RAM cache with local storage would be a big change and we can expect weird browser-specific issues
- Using embedded record would require an id 
- ember-data-model-fragments could work but is in beta and adding a dependency for that specific use case is a bit overkill

The cleanest solution I could think of would be to have a separate model for each setting screen, but this seems a big refactoring that needs to be discussed.

jbenezech commented on ghost/admin/app/adapters/setting.js:
```diff
@@ -12,6 +13,14 @@ export default class Setting extends ApplicationAdapter {
 
+        // Do not send empty data to the API
+        // This can probably be removed then this is fixed:
+        // https://github.com/TryGhost/Ghost/blob/main/ghost/api-framework/lib/validators/input/all.js#L128
```
Performing a partial update on a `PUT` is not very "RESTful". I would be more inclined in using a `PATCH`, but this method is not used anywhere else in the API, so I guess this was a design decision and, on that route at least, `PUT` does behave like a `PATCH`. A partial update with "nothing changed", although pretty useless, is not a bad request in my opinion.
That being said, I lack background information on the project and the API implementation.

jbenezech commented on ghost/admin/app/controllers/settings/navigation.js:
```diff
@@ -136,6 +136,16 @@ export default class NavigationController extends Controller {
         try {
             yield RSVP.all(validationPromises);
+
+            // If some attributes have been changed, rebuild
+            // the model arrays or changes will not be detected
+            if (this.get('dirtyAttributes')) {
+                this.set('settings.navigation',
+                    [...this.get('settings.navigation')]);
+                this.set('settings.secondaryNavigation',
+                    [...this.get('settings.secondaryNavigation')]);
+            }
+
```
> This is ugly

Yep... quick fix :mask:

kevinansfield commented:
Heads up that I've pushed a rebase as some other changes in `main` caused conflicts with the PR

kevinansfield committed:
Updated component tests for moved nav item components

refs https://github.com/TryGhost/Ghost/commit/48af4ab3a1b679b69abe56b96ebf5489eff17310

- updated test locations and component names
- updated tests to use angle bracket syntax to match component usage

ErisDS commented:
Hi @jbenezech thanks so much for this PR! Apologies for the delay - we had a bit of an influx - but it has now been merged 🎉 and will appear in the next release of Ghost - usually Fridays.

I'm not sure if you found this through hacktoberfest, but I've added the accepted label to this PR to make sure it counts. I see you have another PR open which I'll take a look at next!

jbenezech commented:
@ErisDS 
No worries, thanls for merging that and the others :)
Nothing to do with hacktoberfest, more the automatic PR message you add. Working on it ;)
